File: MayTheFourth_C#/Sprites/Sprite.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Audio;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.GamerServices;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using Microsoft.Xna.Framework.Media;

namespace MayTheFourth.Sprites {
    public abstract class Sprite : DrawableGameComponent {
        public Game1 game;

        // Graphics
        public Texture2D texture;
        public Rectangle drawBox;
        public Color color = Color.White;
        public float scale = 1f;
        public float rotation = 0f;

        public Physics physics = new Physics();

        public float vel_max; // = 1f;
        public float acc_max; // = 1f;

        public Sprite(Game1 game) : base(game) {
            this.game = game;
        }

        public override void Initialize() {
            base.Initialize();
        }

        protected override void LoadContent() {
            base.LoadContent();
        }

        public override void Update(GameTime gameTime) {
            physics.Verlet((float) gameTime.ElapsedGameTime.Milliseconds / 16f);
            rotation = physics.yaw_pos;

            drawBox = CreateRectangle(texture, physics.pos, scale);

            base.Update(gameTime);
        }

        public override void Draw(GameTime gameTime) {
            SpriteBatch spriteBatch = game.spriteBatch;

            Vector2 origin = GetOrigin(texture);
            spriteBatch.Draw(texture, drawBox, null, color, rotation, origin, SpriteEffects.None, 0f);

            base.Draw(gameTime);
        }

        public void Forward(int dir = 1) {
            if (physics.vel.Length() < vel_max) {
                physics.acc.X = (float) (Math.Sign(dir) * acc_max * Math.Cos(physics.yaw_pos));
                physics.acc.Y = (float) (Math.Sign(dir) * acc_max * Math.Sin(physics.yaw_pos));
            }
            else {
                physics.acc.X = 0;
                physics.acc.Y = 0;
            }
        }

        public void TurnYaw(float ang = 1) {
            physics.yaw_vel = (float) (MathHelper.ToRadians(ang));
        }

        public void Roll(float ang = 1) {
            physics.acc.X = (float) (ang * Math.Cos(physics.yaw_pos + MathHelper.PiOver2));
            physics.acc.Y = (float) (ang * Math.Sin(physics.yaw_pos + MathHelper.PiOver2));
        }

        public void Friction() {
            int dir = Math.Sign(physics.vel.X);
            
            if (Math.Abs(physics.vel.Length()) > 0) {
                physics.vel *= 0.9f;
            }

            if (Math.Abs(physics.yaw_vel) > 0) {
                physics.yaw_vel *= 0.9f;
            }
        }

        public Rectangle CreateRectangle(Texture2D texture, Vector2 pos, float scale = 1f) {
            int x = (int) pos.X;
            int y = (int) pos.Y;
            int width = (int) (texture.Width * scale);
            int height = (int) (texture.Height * scale);
            Rectangle rect = new Rectangle(x, y, width, height);
            return rect;
        }

        public Vector2 GetOrigin(Texture2D texture) {
            return new Vector2(texture.Width / 2, texture.Height / 2);
        }
    }
}

File: source/Sprite.js
var Sprite = function (game) {
	// Graphics
	// /*public Image*/ texture;
	// /*public Rectangle*/ drawBox;
	/*public Color*/ this.color = "#ffffff";
	/*public float*/ this.scale = 1.0;
	/*public float*/ this.rotation = 0.0;

	/*public Physics*/ this.physics = new Physics();

	/*public float*/ this.vel_max= 1.0;
	/*public float*/ this.acc_max= 1.0;

	/*private Game*/ this.game = game;
}

Sprite.prototype.init = function() {
}

Sprite.prototype.loadContent = function(contentManager) {
}

Sprite.prototype.update = function(dt) {
	this.physics.verlet(dt);
	this.rotation = this.physics.yaw_pos;

	this.drawBox = this.createRectangle(this.texture, this.physics.pos, this.scale);
}

Sprite.prototype.draw = function(/*CanvasContext*/ surface) {
	// TODO: rotation, color, etc
	// spriteBatch.Draw(texture, drawBox, null, color, rotation, origin, SpriteEffects.None, 0f);
	surface.save(); // push
	// Translate around origin
	surface.translate(this.drawBox.x + this.drawBox.width / 2, this.drawBox.y + this.drawBox.height / 2);
	surface.rotate(this.rotation);
	surface.drawImage(this.texture, -this.drawBox.width / 2, -this.drawBox.height / 2, this.drawBox.width, this.drawBox.height);
	surface.restore(); // pop
}

Sprite.prototype.forward = function(dir) {
	dir = dir || 1;

	if (this.physics.vel.magnitude() < this.vel_max) {
		this.physics.acc.x = /*(float)*/ (Math.sign(dir) * this.acc_max * Math.cos(this.physics.yaw_pos));
		this.physics.acc.y = /*(float)*/ (Math.sign(dir) * this.acc_max * Math.sin(this.physics.yaw_pos));
	}
	else {
		this.physics.acc.x = 0;
		this.physics.acc.y = 0;
	}
}

Sprite.prototype.turnYaw = function(ang) {
	ang = ang || 1;

	this.physics.yaw_vel = /*(float)*/ (MathHelper.toRadians(ang));
}

Sprite.prototype.roll = function(ang) {
	ang = ang || 1;

	this.physics.acc.x = /*(float)*/ (ang * Math.cos(this.physics.yaw_pos + MathHelper.PiOver2));
	this.physics.acc.y = /*(float)*/ (ang * Math.sin(this.physics.yaw_pos + MathHelper.PiOver2));
}

Sprite.prototype.friction = function() {
	/*int*/ var dir = Math.sign(this.physics.vel.x);

	if (Math.abs(this.physics.vel.magnitude()) > 0) {
		this.physics.vel = this.physics.vel.mulScalar(0.9);
	}

	if (Math.abs(this.physics.yaw_vel) > 0) {
		this.physics.yaw_vel *= 0.9;
	}
}

/*Rectangle*/ Sprite.prototype.createRectangle = function(/*Texture2D*/ texture, /*Vector2*/ pos, /*float*/ scale) {
	scale = scale || 1.0;

	var origin = this.getOrigin(texture);
	/*int*/ var x = Math.floor(pos.x - origin.x * scale);
	/*int*/ var y = Math.floor(pos.y - origin.y * scale);
	/*int*/ var width = Math.floor(texture.width * scale);
	/*int*/ var height = Math.floor(texture.height * scale);
	/*Rectangle*/ var rect = new Rectangle(x, y, width, height);

	// TODO: this probably should go in init
	this.width = width;
	this.height = height;

	return rect;
}

// Default origin point is center of image
/*Vector2*/ Sprite.prototype.getOrigin = function(/*Texture2D*/ texture) {
	return new Vector2(texture.width / 2, texture.height / 2);
}
